Sources/TSCBasic/Thread.swift

```diff
@@ -85,3 +95,60 @@ final private class ThreadImpl: Foundation.Thread {
 // Thread on Linux supports closure so just use it directly.
 typealias ThreadImpl = Foundation.Thread
 #endif
+
+protocol Defaultable {
+    static var defaultValue: Self { get }
+}
+
+extension Optional: Defaultable {
+    static var defaultValue: Optional<Wrapped> { .none }
+}
+
+/// `ThreadLocal` properties are  thread-specific. Every thread has its own instance of the wrapped property.
+@propertyWrapper final class ThreadLocal<Value: Defaultable> {
+    private var storage: NSMutableDictionary { ThreadImpl.current.threadDictionary }
+    private let key = UUID().uuidString
+
+    var wrappedValue: Value {
+        get {
+            if let value = storage[key] as? Value {
+                return value
+            } else {
+                let value = Value.defaultValue
+                storage[key] = value
+                return value
+            }
+        }
+        set { storage[key] = newValue }
+    }
+
+    init(wrappedValue: Value) {
+        self.wrappedValue = wrappedValue
+    }
+}
+
+/// Automatically closes the wrapped file descriptor on deinit.
+@propertyWrapper final class AutoClosing: NSObject, Defaultable {
```

Comment:
As with the other property wrapper, I think this could have a lot of utility but think it would be better to discuss as a follow-on PR to simplify the code.  I'm also hopeful that things like this will be able to come from SwiftSystem so that ToolSupportCore doesn't need to implement it (https://swift.org/blog/swift-system/).

Reply:
+1 for proposing such generalizations as separate followup PR so can be discussed individually